Node Conformance Test: Add system verification

Random-Liu opened the pull request:
For #30122 and #29081.

This PR introduces system verification test in node e2e and conformance test. It will run before the real test. Once the system verification fails, the test will just fail. The output of the system verification is like this:

```
I0909 23:33:20.622122    2717 validators.go:45] Validating os...
OS: Linux
I0909 23:33:20.623274    2717 validators.go:45] Validating kernel...
I0909 23:33:20.624037    2717 kernel_validator.go:79] Validating kernel version
KERNEL_VERSION: 3.16.0-4-amd64
I0909 23:33:20.624146    2717 kernel_validator.go:93] Validating kernel config
CONFIG_NAMESPACES: enabled
CONFIG_NET_NS: enabled
CONFIG_PID_NS: enabled
CONFIG_IPC_NS: enabled
CONFIG_UTS_NS: enabled
CONFIG_CGROUPS: enabled
CONFIG_CGROUP_CPUACCT: enabled
CONFIG_CGROUP_DEVICE: enabled
CONFIG_CGROUP_FREEZER: enabled
CONFIG_CGROUP_SCHED: enabled
CONFIG_CPUSETS: enabled
CONFIG_MEMCG: enabled
I0909 23:33:20.679328    2717 validators.go:45] Validating cgroups...
CGROUPS_CPU: enabled
CGROUPS_CPUACCT: enabled
CGROUPS_CPUSET: enabled
CGROUPS_DEVICES: enabled
CGROUPS_FREEZER: enabled
CGROUPS_MEMORY: enabled
I0909 23:33:20.679454    2717 validators.go:45] Validating docker...
DOCKER_GRAPH_DRIVER: aufs
```

It verifies the system following a predefined `SysSpec`:

``` go
// DefaultSysSpec is the default SysSpec.
 var DefaultSysSpec = SysSpec{
    OS:            "Linux",
    KernelVersion: []string{`3\.[1-9][0-9].*`, `4\..*`}, // Requires 3.10+ or 4+
    // TODO(random-liu): Add more config
    KernelConfig: KernelConfig{
        Required: []string{
            "NAMESPACES", "NET_NS", "PID_NS", "IPC_NS", "UTS_NS",
            "CGROUPS", "CGROUP_CPUACCT", "CGROUP_DEVICE", "CGROUP_FREEZER",
            "CGROUP_SCHED", "CPUSETS", "MEMCG",
        },
        Forbidden: []string{},
    },
    Cgroups: []string{"cpu", "cpuacct", "cpuset", "devices", "freezer", "memory"},
    RuntimeSpec: RuntimeSpec{
        DockerSpec: &DockerSpec{
            Version: []string{`1\.(9|\d{2,})\..*`}, // Requires 1.9+
            GraphDriver: []string{"aufs", "overlay", "devicemapper"},
        },
    },
 }
```

Currently, it only supports:
- Kernel validation: version validation and kernel configuration validation
- Cgroup validation: validating whether required cgroups subsystems are enabled.
- Runtime Validation: currently, only validates docker graph driver.

The validating framework is ready. The specific validation items could be added over time.

@dchen1107 
/cc @kubernetes/sig-node

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/32427)

<!-- Reviewable:end -->

dchen1107 commented:
Please remove the dependency of #31093 from this one. I think this one should be in first before the packaging solution.

Random-Liu committed:
Add system verification.

Random-Liu commented:
@dchen1107 Removed the dependency.

Random-Liu commented:
@k8s-bot test this issue #IGNORE.

dchen1107 commented on test/e2e_node/system/types.go:
```diff
@@ -0,0 +1,78 @@
+	Cgroups: []string{"cpu", "cpuacct", "cpuset", "devices", "freezer", "memory"},
+	RuntimeSpec: RuntimeSpec{
+		DockerSpec: &DockerSpec{
+			GraphDriver: []string{"aufs", "overlay", "overlay2", "devicemapper"},
```
I don't think we validate overlay2 in any format. It might work, but we need to kick out the qualification process before we claim we support that.

dchen1107 replied:
Ok, I just realized that when overlayfs moved to kernel tree, the upstream called it overlay2. What you are referring to graphdrivers specifically named by docker. 

From https://docs.docker.com/engine/userguide/storagedriver/overlayfs-driver/ it says "For the overlay2 driver, the version of your kernel must be 4.0 or newer" which should be fine on GCI & CoreOS cases (cc/ @euank @justinsb @philips). But should we tight our validation here against kernel version too? 

Also I just realized that we don't know how the docker upgrade will work from 'overlay' to 'overlay2'. We didn't validate that part when we validate docker 1.12.

dchen1107 commented on test/e2e_node/system/types.go:
```diff
@@ -0,0 +1,78 @@
+		Forbidden: []string{},
+	},
+	Cgroups: []string{"cpu", "cpuacct", "cpuset", "devices", "freezer", "memory"},
+	RuntimeSpec: RuntimeSpec{
```
For runtime, we should check minimal support version too. Today, we did this check in kubelet core, but I think it makes sense to all those validation pieces switch to this validation library.

Random-Liu replied:
Yeah, I also realized that just now. Will do that. :)

Random-Liu commented:
@dchen1107 Addressed comments.
Added docker version check (>= 1.9.*)
Removed `overlay2` from graph driver list and add a TODO to validate `overlay2`.

yujuhong commented on test/e2e_node/system/validators.go:
```diff
@@ -0,0 +1,52 @@
+	for _, v := range validators {
+		glog.Infof("Validating %s...", v.Name())
+		err := v.Validate(spec)
+		if err != nil {
```
nit: `errors.NewAggregate()` actually checks to see if the errors in the list are nil or not, so you don't have to check here.

Random-Liu replied:
ACK.

yujuhong commented on test/e2e_node/system/kernel_validator.go:
```diff
@@ -0,0 +1,222 @@
+	k.kernelRelease = strings.TrimSpace(string(out))
+	var errs []error
+	if err := k.validateKernelVersion(spec.KernelVersion); err != nil {
+		errs = append(errs, err)
```
nit: same here; no need to check the error before appending

Random-Liu replied:
ACK. :)

Random-Liu commented:
@dchen1107 PTAL~

luxas commented:
@dchen1107 @Random-Liu @yujuhong What's the ETA on this?
It would be nice to include it in kubeadm as well...

errordeveloper commented:
As we'd like to use this in `kubeadm`, it'd be great if output was a little prettier! I'd be happy to help, if you want 👍

dchen1107 commented:
Here are couple of things missing in current validation tests as the minimal requirements for kernel:

1) Network:
  - CONFIG_INET
  - CONFIG_IP_NF_TARGET_REDIRECT or CONFIG_NETFILTER_XT_TARGET_REDIRECT, and 
CONFIG_NETFILTER_XT_MATCH_COMMENT: Otherwise kube-proxy won't work fine. 

2) Storage and filesystem: 
 - CONFIG_EXT4_FS, CONFIG_PROC_FS
 - CONFIG_OVERLAY or CONFIG_AUFS or devicemapper (? @derekwaynecarr on depended kernel module & config)
 
 3) Resource isolation and management:
  - I saw you already includes all required namespace and cgroups. checked. 

There are more required based on the feature set. For example, if the user uses Gluster volume, Kernel has to configured with CONFIG_FUSE; if the user uses Calico networking, CONFIG_IP_SET_HASH_IP has to be set. We can extend this in next version for those as a warning with detail information, not as the minimal requirement here.

Random-Liu commented:
@errordeveloper @luxas Sorry for the delay.

We'll finish this soon.

@errordeveloper Feel free to send PR to make this prettier! That's really appreciated! :)

luxas commented:
It would be great if we could get this PR in before the code freeze. Thanks!

Random-Liu commented:
@dchen1107 Addressed comments. Here is the new output:
```

OS: Linux
KERNEL_VERSION: 4.4.21+
CONFIG_NAMESPACES: enabled
CONFIG_NET_NS: enabled
CONFIG_PID_NS: enabled
CONFIG_IPC_NS: enabled
CONFIG_UTS_NS: enabled
CONFIG_CGROUPS: enabled
CONFIG_CGROUP_CPUACCT: enabled
CONFIG_CGROUP_DEVICE: enabled
CONFIG_CGROUP_FREEZER: enabled
CONFIG_CGROUP_SCHED: enabled
CONFIG_CPUSETS: enabled
CONFIG_MEMCG: enabled
CONFIG_INET: enabled
CONFIG_EXT4_FS: enabled
CONFIG_PROC_FS: enabled
CONFIG_NETFILTER_XT_TARGET_REDIRECT: enabled (as module)
CONFIG_NETFILTER_XT_MATCH_COMMENT: enabled
CONFIG_OVERLAY_FS: enabled
CONFIG_AUFS_FS: not set - Required for aufs.
CONFIG_BLK_DEV_DM: enabled
CGROUPS_CPU: enabled
CGROUPS_CPUACCT: enabled
CGROUPS_CPUSET: enabled
CGROUPS_DEVICES: enabled
CGROUPS_FREEZER: enabled
CGROUPS_MEMORY: enabled
DOCKER_VERSION: 1.11.2
DOCKER_GRAPH_DRIVER: overlay
PASS
```

Random-Liu committed:
Update bazel.

Random-Liu committed:
Add unit test for system verification

Random-Liu commented:
@dchen1107 I've included unit test from https://github.com/kubernetes/kubernetes/pull/32454 into this PR so as to get the change merged soon.

dchen1107 commented:
@k8s-bot gci gce e2e test this

dchen1107 commented:
LGTM

Random-Liu commented:
Apply LGTM based on https://github.com/kubernetes/kubernetes/pull/32427#issuecomment-258487033.
